### PR TITLE
Xworkflow 1 for XkOps

### PR DESCRIPTION
**Pull request by SyedMurtazaHassan46**
# Xworkflow 1

## Issue

[Issue no 57](https://github.com/XgridInc/xkops/issues/57)

## Description

- Created a go script for kubecost to fetch unclaimed volumes from the cluster and save it in the mongodb statefulset
- Created dockerfile for the kubecost script
- Created cronjob using the docker image of the kubecost script dockerfile
- Created headless service, storage class and statefulset for mongodb
- Created example unclaimed volume without claim for testing.

## Checklist

- [ ] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `docs` directory)
- [x] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes (if applicable)
- [x] Every function, interface, class has a comment describing what it does and input/output parameters

### CODE REVIEW

File: xWorkflows/Xworkflow1/Readme.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    ```
+    rs.initiate({
+    _id: "xworkflowReplSet",
+    members: [
+        { _id: 0, host: "xworkflow-mongodb-0.xworkflow-mongodb.default.svc.cluster.local:27017" },
+        { _id: 1, host: "xworkflow-mongodb-1.xworkflow-mongodb.default.svc.cluster.local:27017" },
+        { _id: 2, host: "xworkflow-mongodb-2.xworkflow-mongodb.default.svc.cluster.local:27017" }
+    ]
+    });
+    ```
```
Comment (alikhanxgrid) [inline xWorkflows/Xworkflow1/Readme.md:51]: cant we do this without exec ? Like when deploying mongodb shouldn't this be a default behavior ?

File: xWorkflows/Xworkflow1/backend/app.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+from pymongo import MongoClient
+from flask import Flask, jsonify, request
+
+app = Flask(__name__)
+
+# MongoDB connection URI and database/collection names
+mongoURI = "mongodb://xworkflow-mongodb-0.xworkflow-mongodb.default.svc.cluster.local:27017," \
+           "xworkflow-mongodb-1.xworkflow-mongodb.default.svc.cluster.local:27017," \
+           "xworkflow-mongodb-2.xworkflow-mongodb.default.svc.cluster.local:27017/?replicaSet=xworkflowReplSet"
+dbName = "xworkflow-db"
```
Comment (umerxgrid) [inline xWorkflows/Xworkflow1/backend/app.py:11]: Same here, using env var

File: xWorkflows/Xworkflow1/go/kubecost-event/main.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+	"fmt"
+	"io"
+	"net/http"
+
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// Kubecost API endpoint URL (replace with your actual URL)
+const kubecostAPIURL = "http://kubecost-cost-analyzer.kubecost.svc.cluster.local:9003/allPersistentVolumes"
```
Comment (umerxgrid) [inline xWorkflows/Xworkflow1/go/kubecost-event/main.go:15]: Using env var, and providing variable values using env var in pod.spec yaml file

File: xWorkflows/Xworkflow1/go/kubecost-event/main.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+			}
+			return fmt.Errorf("error inserting document into MongoDB: %w", err)
+		}
+	}
+	return nil
+}
+
+func main() {
+	// MongoDB connection URI
+	mongoURI := "mongodb://xworkflow-mongodb-0.xworkflow-mongodb.default.svc.cluster.local:27017,xworkflow-mongodb-1.xworkflow-mongodb.default.svc.cluster.local:27017,xworkflow-mongodb-2.xworkflow-mongodb.default.svc.cluster.local:27017/?replicaSet=xworkflowReplSet"
```
Comment (umerxgrid) [inline xWorkflows/Xworkflow1/go/kubecost-event/main.go:114]: Suggestion: using env var here and passing value using pod.spec container env variables

**Review by umerxgrid — APPROVED** [review]
+1 with some minor suggestions e.g using env vars in go and python files

**Review by alikhanxgrid — APPROVED** [review]
+1

File: xWorkflows/Xworkflow1/manifests/unclaimedvolumebackend/service.yaml [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+metadata:
+  name: unclaimed-pv-api-service
+spec:
+  selector:
+    app: unclaimed-pv-api
+  ports:
+  - protocol: TCP
+    port: 5000
+    targetPort: 5000
+  type: NodePort
```
Comment (HamzaXgrid) [inline xWorkflows/Xworkflow1/manifests/unclaimedvolumebackend/service.yaml:12]: why we are not using clusterIP

**Review by HamzaXgrid — APPROVED** [review]
+1

File: xWorkflows/Xworkflow1/manifests/unclaimedvolumebackend/service.yaml (hunk 5 above)
Comment (SyedMurtazaHassan46) [reply xWorkflows/Xworkflow1/manifests/unclaimedvolumebackend/service.yaml:12]: For now, we need to test with the backend with Postman, that is why this is used.

File: xWorkflows/Xworkflow1/Readme.md (hunk 1 above)
Comment (SyedMurtazaHassan46) [reply xWorkflows/Xworkflow1/Readme.md:51]: this is optional if you use multiple pods in MongoDB, we will update this when working on db consolidation if we need this.

File: xWorkflows/Xworkflow1/go/kubecost-event/main.go (hunk 3 above)
Comment (SyedMurtazaHassan46) [reply xWorkflows/Xworkflow1/go/kubecost-event/main.go]: added env vars for these values

File: xWorkflows/Xworkflow1/go/kubecost-event/main.go (hunk 4 above)
Comment (SyedMurtazaHassan46) [reply xWorkflows/Xworkflow1/go/kubecost-event/main.go]: added env vars for these values

File: xWorkflows/Xworkflow1/backend/app.py (hunk 2 above)
Comment (SyedMurtazaHassan46) [reply xWorkflows/Xworkflow1/backend/app.py]: added env vars for these values

File: xWorkflows/Xworkflow1/backend/app.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
+import requests
+from pymongo import MongoClient
+from flask import Flask, jsonify, request
+
+app = Flask(__name__)
+
+# MongoDB connection URI and database/collection names
+mongoURI = "mongodb://xworkflow-mongodb-0.xworkflow-mongodb.default.svc.cluster.local:27017," \
```
[the remark this answers is not in the document]
Comment (SyedMurtazaHassan46) [reply xWorkflows/Xworkflow1/backend/app.py]: added env var for these values

**Pull request merged**